realm/realm-library/src/main/cpp/io_realm_internal_OsObject.cpp

```diff
@@ -122,8 +122,10 @@ struct ChangeCallback {
         m_wrapper->m_row_object_weak_ref.call_with_local_ref(env, [&](JNIEnv*, jobject row_obj) {
             static JavaMethod notify_change_listeners(env, row_obj, "notifyChangeListeners",
                                                       "([Ljava/lang/String;)V");
-            env->CallVoidMethod(row_obj, notify_change_listeners, m_field_names_array);
+            env->CallVoidMethod(row_obj, notify_change_listeners, m_deleted ? nullptr : m_field_names_array);
         });
+        m_field_names_array = nullptr;
```

Comment:
it could be set in `before()`. The code will become clearer when implementing the old/new value.